### PR TITLE
Move to junit5 + hamcrest matchers to address reflection warning

### PR DESCRIPTION
**Pull request by kn100**
Compiling with anything more modern than Java 7 threw a warning about Illegal reflection to do with junit4s use of a code generation library called CGLIB. I upgraded a few dependencies and fixed the tests.

### CODE REVIEW

**Comment by kn100** [comment]
Why?

This will allow us to claim compatibility with the latest versions of java without ugly compiler warnings.

**Comment by WillSewell** [comment]
It would be nice if we had an official policy for the level of support we provide. I suggest we go with versions that have "Premier Support" from Oracle: https://www.oracle.com/technetwork/java/java-se-support-roadmap.html. The support policy for OpenJDK has shorter durations: https://adoptopenjdk.net/support.html#roadmap. WDYT @jameshfisher?

File: src/test/java/com/pusher/rest/PusherHttpTest.java [hunk 1]
```diff
@@ -33,7 +33,7 @@
 
     private Pusher p;
 
-    @Before
+    @BeforeEach
```
Comment (WillSewell) [inline src/test/java/com/pusher/rest/PusherHttpTest.java]: Why did you change this from `Before` to `BeforeEach`?

Comment (kn100) [reply src/test/java/com/pusher/rest/PusherHttpTest.java]: Because Before doesn't exist in Junit5 anymore.

Comment (WillSewell) [reply src/test/java/com/pusher/rest/PusherHttpTest.java]: Ah, I see, it was renamed https://stackoverflow.com/a/45037910/1018290.

**Comment by jameshfisher** [comment]
Yes, I think this is a reasonable policy
